src/librustc/mir/visit.rs

```diff
@@ -120,181 +120,182 @@ macro_rules! yrt { // try backwards
 
 macro_rules! make_mir_visitor {
     ($visitor_trait_name:ident, $($mutability:ident)?) => {
-        pub trait $visitor_trait_name<'tcx> {
+        pub trait $visitor_trait_name<'tcx, R: TryHarder<Ok = ()> = ()> {
```

Comment:
Can we use `Result<T, E>` with an associated error type that is sometimes `!` instead? I guess that requires `Ok` wrapping, but seems overall simpler, and we could continue to use `?`?

Reply:
That would require modifying all implementers of `Visitor` right? I didn't want to do this until there was some interest in it. I agree that it would be nicer though.

Reply:
Yes, it would require that. We could of course do it not with an associated type but with a type parameter, just as you did here 

```rust
trait Visitor<'tcx, E = !> { ... }
```

that may be better anyway